presto-main/src/main/java/com/facebook/presto/sql/planner/SubqueryPlanner.java

```diff
@@ -223,18 +240,137 @@ private PlanBuilder appendSubqueryApplyNode(PlanBuilder subPlan, Expression subq
+        return correlation.build();
+    }
+
     private RelationPlan createRelationPlan(Query subquery)
     {
         return new RelationPlanner(analysis, symbolAllocator, idAllocator, metadata, session)
                 .process(subquery, null);
     }
+
+    private Set<Expression> extractMissingExpressions(PlanNode planNode)
+    {
+        return ExpressionExtractor.extractExpressions(planNode).stream()
+                .flatMap(expression -> extractColumnReferences(expression, analysis.getColumnReferences()).stream())
+                .collect(toImmutableSet());
+    }
+
+    private static Set<Expression> extractColumnReferences(Expression expression, Set<Expression> columnReferences)
+    {
+        ImmutableSet.Builder<Expression> expressionColumnReferences = ImmutableSet.builder();
+        new ColumnReferencesExtractor(columnReferences).process(expression, expressionColumnReferences);
+        return expressionColumnReferences.build();
+    }
+
+    private PlanNode replaceExpressionsWithSymbols(PlanNode planNode, Map<Expression, Symbol> mapping)
+    {
+        if (mapping.isEmpty()) {
+            return planNode;
+        }
+
+        Map<Expression, Expression> expressionMapping = mapping.entrySet().stream()
+                .collect(toImmutableMap(Map.Entry::getKey, e -> e.getValue().toSymbolReference()));
+        return SimplePlanRewriter.rewriteWith(new ExpressionReplacer(idAllocator, expressionMapping), planNode, null);
+    }
+
+    private static class ColumnReferencesExtractor
+            extends DefaultExpressionTraversalVisitor<Void, ImmutableSet.Builder<Expression>>
+    {
+        private final Set<Expression> columnReferences;
+
+        private ColumnReferencesExtractor(Set<Expression> columnReferences)
+        {
+            this.columnReferences = requireNonNull(columnReferences, "columnReferences is null");
+        }
+
+        @Override
+        protected Void visitDereferenceExpression(DereferenceExpression node, ImmutableSet.Builder<Expression> builder)
+        {
+            if (columnReferences.contains(node)) {
+                builder.add(node);
+            }
+            else {
+                process(node.getBase(), builder);
+            }
+            return null;
+        }
+
+        @Override
+        protected Void visitQualifiedNameReference(QualifiedNameReference node, ImmutableSet.Builder<Expression> builder)
+        {
+            builder.add(node);
+            return null;
+        }
+    }
+
+    private static class ExpressionReplacer
+            extends SimplePlanRewriter<Void>
+    {
+        private final PlanNodeIdAllocator idAllocator;
+        private final Map<Expression, Expression> mapping;
+
+        public ExpressionReplacer(PlanNodeIdAllocator idAllocator, Map<Expression, Expression> mapping)
+        {
+            this.idAllocator = requireNonNull(idAllocator, "idAllocator is null");
+            this.mapping = requireNonNull(mapping, "mapping is null");
+        }
+
```

Comment:
IMO all nodes are supported. Default behavior is to visit children nodes (SimplePlanRewriter). 

However, the problem here is that ExpressionReplacer can miss some nodes which requires expression to rewritten (case of some visit* missing). As all such method has to be listed (handled) here explicitly. But this is a common problem with visitor pattern.